src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,615 @@
+         } else {
+            // Assume the document is part of a Match Expression, like:
+            // {age: {$gt: 5}}
+            arg_type = MATCH_EXPRESSION;
+         }
+      }
+
+      operator_value_t op;
+      switch (arg_type) {
+      case AGGREGATE_EXPRESSION:
+         if (!parse_aggregate_expression (in, doc, &op, status)) {
+            goto fail;
+         }
+         break;
+      case MATCH_EXPRESSION:
+         if (!parse_match_expression (in, doc, &op, status)) {
+            goto fail;
+         }
+         break;
+      case UNKNOWN:
+      default:
+         ERR_WITH_BSON (in, "%s", "unexpected unknown expression type");
+         goto fail;
+      }
+
+      switch (op.op_type) {
+      case FLE2RangeOperator_kGt:
+      case FLE2RangeOperator_kGte:
+         if (spec->lower.set) {
+            ERR_WITH_BSON (in, "unexpected duplicate bound %s", op.op_type_str);
+            goto fail;
+         }
+         spec->lower.set = true;
+         spec->lower.value = op.value;
+         spec->lower.included = op.op_type == FLE2RangeOperator_kGte;
+         break;
+      case FLE2RangeOperator_kLt:
+      case FLE2RangeOperator_kLte:
+         if (spec->upper.set) {
+            ERR_WITH_BSON (in, "unexpected duplicate bound %s", op.op_type_str);
+            goto fail;
+         }
+         spec->upper.set = true;
+         spec->upper.value = op.value;
+         spec->upper.included = op.op_type == FLE2RangeOperator_kLte;
+         break;
+      case FLE2RangeOperator_kNone:
+      default:
+         ERR_WITH_BSON (in, "unsupported operator type %s", op.op_type_str);
+         goto fail;
+         break;
+      }
+
+      if (spec->field) {
+         if (0 != strcmp (spec->field, op.field)) {
+            ERR_WITH_BSON (
+               in,
+               "unexpected field mismatch. Expected all fields to be "
+               "%s, but got %s",
+               spec->field,
+               op.field);
+            goto fail;
+         }
+      }
+
+      spec->nOps++;
+
+      if (spec->nOps == 1) {
+         spec->firstOp = op.op_type;
+      } else if (spec->nOps == 2) {
+         spec->secondOp = op.op_type;
+      } else {
+         ERR_WITH_BSON (
+            in, "expected 1 or 2 operators, got > 2: %s", op.op_type_str);
+         goto fail;
+      }
+
+      spec->field = op.field;
+   }
+
+   ok = true;
+fail:
+   return ok;
+}
+
+// mc_makeRangeFindPlaceholder creates a placeholder to be consumed by
+// libmongocrypt for encryption.
+bool
+mc_makeRangeFindPlaceholder (mc_makeRangeFindPlaceholder_args_t args,
+                             _mongocrypt_buffer_t *out,
+                             mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bool ok = false;
+   bson_t *edgesInfo = bson_new ();
+   bson_t *v = bson_new ();
+   bson_t *p = bson_new ();
+   _mongocrypt_buffer_init (out);
+
+#define TRY(stmt)                                          \
```

Comment:
Can `TRY` be replaced with `BSON_ASSERT` instead?

Question also applies to other instances in this file.

Reply:
Ditto of other comment regarding input validation.